Watch tickbox

aliceh75 opened the pull request:
- Add a watch tickbox on the detail page;
- Use an html entity for the tickbox rather than an input tag, as the latter approach fails on some browsers (at least in chromium)
- Stop the tickbox in the list table cell from expending it's clickable area to the whole cell.

aliceh75 committed:
Add 'watch' tickbox on the item detail page

aliceh75 committed:
Add toggle and watch class to the watch tickbox on the list page to allow for styling. Ensure toggle tickboxes click area on the list page don't stretch to the whole table cell, as that is confusing. Show the cursor when hovering above a toggle tickbox for consisency.

aliceh75 committed:
html doesn't allow for input inside anchor tags, and this fails on some browsers. Replace it with a utf8 tickbox, since it's only used for the look anyway.

birdsarah commented on django/econsensus/publicweb/static/css/styles.css:
```diff
@@ -111,6 +111,16 @@ a.button {
 	top:0.2em;
 }
 
+a.toggle {
```
my understanding from https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Writing_efficient_CSS is that it's not good practice to qualify a class rules with a tag name. Is there a specific reason why `.toggle` wouldn't do here?

birdsarah commented on django/econsensus/publicweb/static/css/styles.css:
```diff
@@ -441,6 +451,11 @@ input {
 	color: #000;
 }
 
+.summary-list td a.toggle {
```
1) As above

2) I'm wondering if there's a more efficient way of doing this but not too bothered.

birdsarah commented on django/econsensus/publicweb/templates/decision_list.html:
```diff
@@ -66,8 +66,9 @@
-                                        <input type="checkbox"{% if request.user|is_watching:object %} checked="True"{% endif %} />
+                                    <a class='toggle watch'
+                                       href="{% if request.user|is_watching:object %}{% url 'remove_watcher' object.id %}{% else %}{% url 'add_watcher' object.id %}{% endif %}?next={{ request.path_info }}">
+                                        {% if request.user|is_watching:object %}&#x2611;{% else %}&#x2610;{% endif %}
```
given that this doesn't have the text "watch" maybe it could have a title attribute for extra prompting (is that sufficient for screenreaders, not sure)

aliceh75 replied:
After some research it turns out that the W3C discourages using title attribute for links. See the warning under: http://www.w3.org/TR/html/dom.html#the-title-attribute

The W3C suggested approach is to hide part of the link text by moving it out of the viewport in a 1px box: http://www.w3.org/TR/2012/NOTE-WCAG20-TECHS-20120103/C7

I realise this is from 2012, though I can't find anything that says otherwise. I'm not particularly fond of this technique - still, if it is the recommended one I guess we should go ahead. What do you think?

Either way, I guess the text should be more descriptive. Something like 'watch/unwatch <item excerpt>'.

aliceh75 replied:
An alternative approach would be to replace the utf8 tickbox with an image with an 'alt' attribute, as this would be read by screen readers. Cleaner than the option above, though it uses more bandwidth.

birdsarah replied:
This is really interesting. Thanks for looking it up. We may also want to contribute this to http://a11yproject.com/patterns/ I can't see it there (I only looked quickly though)

birdsarah commented on django/econsensus/publicweb/tests/view_decision_test.py:
```diff
@@ -46,3 +50,53 @@ def test_load_decision_form(self):
+        elements = doc.cssselect('div#decision_detail a.watch.toggle.unwatched')
+        self.assertEquals(len(elements), 1)
+
+    def test_following_unwatched_watch_toggle_adds_watcher(self):
```
just wanted to say that i really like this test. not entirely sure why. but i do.

birdsarah commented:
made some comments. nothing that would stop me merging, but i think we can keep pushing to be better on econsensus.

birdsarah commented:
probably worth noting that i have only read the code, i haven't downloaded it and run it.

aliceh75 committed:
Avoid tag.class css rules as they are not recomended. Rename 'toggle' to 'toggle-link' as 'toggle' is too generic.

aliceh75 commented:
I have changed the css from `a.toggle` to `.toggle-link` as suggested. (I renamed it `toggle-link` as I felt `toggle` was too generic).

While the diff is outdated, I did not add the 'watch' title in the list view as I'm not sure what the best approach to do this is. See my comments there.

aliceh75 committed:
Sighted users do not need to see table header information repeated in the links of each row of the table. Users who instead jump from link to link may do. W3C does not recomened using the title attribute for this (see http://www.w3.org/TR/html/dom.html#the-title-attribute ), instead they recomend adding text hidden from view by moving it out of the viewport (see http://www.w3.org/TR/2012/NOTE-WCAG20-TECHS-20120103/C7 ). This commit adds a link to markup such information, and implements this for the watch tick box.

aliceh75 commented:
To see what it would be like, I have committed an implementation of the W3C approach for adding the additional 'watch ...' text in the link in the list table. Let me know what you think.

Summary of the issue (as the relevant line diffs are now out of date):
Sarah suggested adding a 'watch' title to the link in the table view, as it only contains a checkbox which is not accessible. However W3C does not recomened this approach (see warning under http://www.w3.org/TR/html/dom.html#the-title-attribute ) and instead recommend moving the text out of the viewport (see  http://www.w3.org/TR/2012/NOTE-WCAG20-TECHS-20120103/C7 ).

aliceh75 commented:
aria-label seems to be well supported, so that's probably a better approach. Will change the last commit to use that instead. https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-label_attribute

birdsarah commented:
Nice! Thanks for looking this all up. This is really great. I'm happy for you to merge when you submit the aria change.
